### PR TITLE
Average pooling clamped divisor should be done on all conditions where the kernel can go out of bounds

### PR DESCRIPTION
**Pull request by ivangarcia44**
In this pull request I added various E2E tests to cover edge cases on the average pooling torch to linalg lowering algorithm. These new tests uncovered various numerical issues that were addressed in this same PR.

One of the issues is the IREE test failure found in https://github.com/llvm/torch-mlir/issues/4079 which triggered this work.

Background:

In the most common case, the divisor for the average pooling is just the product of kernel dimensions. But with padding, and ceil mode options, some elements need to be discounted from the divisor computation. This change fixes two components of this:

Fix the condition that determines if the divisor is just the product of kernel dimensions or the clamped divisor comptutation.
Add missing isCountIncludePad logic divisor computation algorithm and reversal of kernel/stride/padding parameters element order. Both were missing from the first generalization change.

@vivekkhandelwal1
@silvasean
@zjgarvey
@ramiro050
@JianzheXiao
@AmosLewis
@rsuderman
@nirvedhmeshram
@sahas3
@Hanumanth04
@dixinzhou
@rafaelubalmw

### CODE REVIEW

File: test/Conversion/TorchToLinalg/pooling.mlir [hunk 1]
```diff
@@ -225,3 +225,31 @@ func.func @forward_avg_pool1d_countincludepad_false(%arg0: !torch.vtensor<[1,512
   %3 = torch.aten.avg_pool1d %arg0, %0, %1, %2, %false, %false : !torch.vtensor<[1,512,10],f32>, !torch.list<int>, !torch.list<int>, !torch.list<int>, !torch.bool, !torch.bool -> !torch.vtensor<[1,512,12],f32>
   return %3 : !torch.vtensor<[1,512,12],f32>
 }
+
+// CHECK-LABEL: func @forward_avgpool_2d_ceil
+func.func @forward_avgpool_2d_ceil(%arg0: !torch.vtensor<[1,1,4,4],f32>) -> !torch.vtensor<[1,1,2,2],f32> attributes {torch.onnx_meta.ir_version = 9 : si64, torch.onnx_meta.opset_version = 19 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
```
Comment (sahas3) [inline test/Conversion/TorchToLinalg/pooling.mlir:230]: Are these `torch.onnx_meta.*` attributes required for the test? If not, please remove these.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 2]
```diff
@@ -2302,3 +2302,92 @@ def MaxUnpool3dModulePad0_basic(module, tu: TestUtils):
     output, indices = pool(input)
 
     module.forward(output, indices)
+
+
+class AvgPool2dCeilNoPadNonUnitaryStridesIreeSwa(torch.nn.Module):
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2307]: Don't these tests fail in the TOSA/Stablehlo CI? The implementation to handle count_include_pad = false is specific to `torch-to-linalg` pass.

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 3]
```diff
@@ -1069,8 +1069,14 @@ std::optional<LogicalResult> ConvertAtenAvgPoolOp<OpTy, PoolingOpTy, Dim>::
 
   constexpr int avgPoolDims = getAvgPoolNumOfDims<OpTy>();
 
-  bool noPadding = llvm::all_of(paddingInts, [](int64_t p) { return p == 0; });
-  if (countIncludePad || noPadding) {
+  bool hasPadding =
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:1072]: I don't see anything in https://pytorch.org/docs/stable/generated/torch.nn.functional.avg_pool2d.html that explains this condition -- can you elaborate what the reference is?

In https://github.com/pytorch/pytorch/blob/1eba9b3aa3c43f86f4a2c807ac8e12c4a7767340/torch/nn/modules/pooling.py#L707 I see a note on when to skip the last window -- but the logic there doesn't match with the condition here.

File: test/Conversion/TorchToLinalg/pooling.mlir (hunk 1 above)
Comment (ivangarcia44) [reply test/Conversion/TorchToLinalg/pooling.mlir:230]: They were not needed. Removed. Thanks

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py (hunk 2 above)
Comment (ivangarcia44) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2307]: They do fail with:
python -m e2e_testing.main --config=fx_importer_stablehlo

I thought I was going to catch it with this: bash build_tools/ci/test_posix.sh nightly (which is passing).

I will update the corresponding filters.

Comment (ivangarcia44) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2307]: Some tests pass in TOSA:

****** Unexpectedly Passed tests - 3 tests
    XPASS - "AvgPool2dCeilNoPadNonUnitaryStridesIreeSwa_basic"
    XPASS - "AvgPool2dCeilNoPadUnitaryStrides_basic"
    XPASS - "AvgPool2dCeilPadNonUnitaryStrides_basic"

I had to remove the filters on those. 

I also had to remove this filter:

Unexpected outcome summary: (fx_importer_stablehlo)

****** Unexpectedly Passed tests - 1 tests
    XPASS - "AvgPool2dCeilNoPadUnitaryStrides_basic"

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 3 above)
Comment (ivangarcia44) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:1133]: @sahas3 The documentation in https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html#torch.nn.AvgPool2d does not give much detail on how the divisor is computed.

I found out through experimentation that PyTorch adds implicit padding not controlled by the count_include_pad parameter. In the most basic case, the divisor is just the product of kernel dimensions. But this total has to be reduced by the number of elements not included which could be due to padding or celing mode rounding.

The formula has 4 input parameters, for a total of 16 possibilities. All 16 possibilities have E2E tests that do non-zero numerical equivalence check between PyTorch and torch-mlir. Most of those 16 combinations are covered by new tests.

Some of them found new SWA’s, and most of them were reproducible without the change in this pull request. For one of them the SWA was related to the actual divisor formula, not the condition written for this pull request. I had to rewrite the divisor computation formula. The variables names are clearer now and I added more details in the comments in the source code.

**Comment by ivangarcia44** [comment]
A new PR is made to resolve the merge conflicts: https://github.com/llvm/torch-mlir/pull/4169

**Comment by ivangarcia44** [comment]
I will do the merge in this PR and close the new one.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 4]
```diff
@@ -2514,3 +2514,494 @@ def MaxUnpool3dModulePad0_basic(module, tu: TestUtils):
...
+            ([1, 1, 4, 4], torch.float32, True),
+        ]
+    )
+    def forward(self, x):
+        return self.ap2d(x)
+
+
+@register_test_case(module_factory=lambda: AvgPool2dCeilNoPadNonUnitaryStridesIreeSwa())
+def AvgPool2dCeilNoPadNonUnitaryStridesIreeSwa_basic(module, tu: TestUtils):
+    module.forward(tu.rand(1, 1, 4, 4, low=-1))
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2547]: Don't need to add "IreeSwa" in the test name.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 5]
```diff
@@ -891,57 +894,125 @@ PoolSizeCalculator<NumOfDims>::PoolSizeCalculator(
...
+    //
+    // indexStartOffset = ceil((kernelSize - 1)/2) - padding
+    //
+    // clampedKernelSize =
+    //   min(outIntIndex * stride + indexStartOffset + floor((kernelSize - 1)/2)
+    //   + 1,
+    //       InputSpatialDimSize + padding) -
+    //   max(outIntIndex * stride + indexStartOffset - ceil((kernelSize - 1)/2),
+    //   -padding)
+    //
```
[the remark this answers is not in the document]
Comment (ivangarcia44) [reply lib/Conversion/TorchToLinalg/Pooling.cpp]: Hi Vivek, the original author of the torch-linalg avg_pool2d lowering (@AmosLewis ) based the 2D implementation of the divisor computation on this PyTorch code: 

https://github.com/pytorch/pytorch/blob/4a6dfbe4806b361c43210dfd56db64c4097c66bb/aten/src/ATen/native/cpu/AvgPoolKernel.cpp#L78

For the new E2E test suite I added, the test AvgPool2dCeilPaddingStridedIncludePadding discovered that there was a numerical mismatch due to the divisor computation. Because of this I wrote a new average pooling divisor formula based on the PyTorch avg_pool2d doc and numerical values I observed. I tried to make the comments, variable names and code clearer than the previous version which helped me to correct the numerical issues I found and make all E2E tests pass.

Taking a second look at the PyTorch code, and the @AmosLewis code, I noticed I missed a detail in the 1D/3D generalization of the algorithm: 1) count_include_pad parameter used in the divisor computation 2) Reversing of kernel/stride/pad parameter elements order. Although the 2nd issue could have been present in the original code. The second issue was found by 3 new E2E tests from this PR with non-uniform values for these parameters.

I updated the PR to use the PyTorch based algorithm with these two fixes and all E2E tests are passing now.

**Comment by ivangarcia44** [comment]
Hi all, this is a reminder to review when you get a chance. Thank you

@vivekkhandelwal1
@JianzheXiao
@AmosLewis
@rsuderman
@nirvedhmeshram)
@sahas3
@Hanumanth04
@dixinzhou
@rafaelubalmw

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 6]
```diff
@@ -915,19 +916,26 @@ Value PoolSizeCalculator<NumOfDims>::getPoolSize(
     // Dim below stands for spatial dimension. Prior to the February 2025
     // change, these variables used "height" and "width" (or "h" and "w")
     // in these intermediate variables instead of "Dim".
+
+    // The average pool properties of kernel size, strides, and padding are
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:920]: Other than the refactoring and variable name changes, this logic appears to be the main change in this PR. Is this documented somewhere? Can you add a link for reference?

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 7]
```diff
@@ -1040,27 +1061,59 @@ LogicalResult ConvertAtenAvgPoolOp<OpTy, PoolingOpTy, Dim>::matchAndRewrite(
...
+  // does not take effect.
+  // The divisor count can be clamped also through the ceil_mode. In this
+  // case, according to the Hout and Wout formula in this page:
+  // https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html#torch.nn.AvgPool2d,
+  // the ceil_mode will round up on the stride division. The round up
+  // will give an extra element that will go out of bounds which PyTorch
+  // adds zero padding in it. It also does not count the implicit zero
+  // padding elements in the divisor, and it is not controlled by the
+  // count_include_pad argument.
+  // But also note that if all strides are 1 there are not fractions to
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:1095]: typo: not -> no

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 8]
```diff
@@ -937,11 +945,15 @@ Value PoolSizeCalculator<NumOfDims>::getPoolSize(
     Value IDim1Clamped = b.createOrFold<arith::MinSIOp>(location, IDim1, IDim);
     Value IDim1_IDim0_Clamped =
         b.createOrFold<arith::SubIOp>(location, IDim1Clamped, IDim0Clamped);
+
+    Value poolSizeDim =
+        !isCountIncludePad
+            ? IDim1_IDim0_Clamped
+            : b.createOrFold<arith::SubIOp>(location, IDim1, IDim0);
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:952]: This is also new change in this PR? What is the reference for this change?

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 9]
```diff
@@ -2514,3 +2514,494 @@ def MaxUnpool3dModulePad0_basic(module, tu: TestUtils):
...
+    def forward(self, x):
+        return self.ap2d(x)
+
+
+@register_test_case(module_factory=lambda: AvgPool2dCeilPaddingStridedIncludePadding())
+def AvgPool2dCeilPaddingStridedIncludePadding_basic(module, tu: TestUtils):
+    module.forward(tu.rand(1, 1, 4, 4, low=-1))
+
+
+class AvgPool2dDiffKernelsStridesNoPadCeilPadNotIncluded(torch.nn.Module):
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2825]: Doesn't this case also cover the similar scenario with non-unit strides? Do we need the other test then to avoid over-testing?

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 7 above)
Comment (ivangarcia44) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:1095]: Fixed thanks

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py (hunk 9 above)
Comment (ivangarcia44) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2825]: Replaced the non-unit stride test with AvgPool2dDiffKernelsStridesNoPadCeilPadNotIncluded. Updated the AvgPool2dDiffKernelsStridesNoPadCeilPadNotIncluded comments to reflect this.

**Comment by ivangarcia44** [comment]
Hi @vivekkhandelwal1,

Could you please take a look at this PR when you have a chance? I’ve addressed all the previous feedback, and GitHub requires your approval before it can be merged. This PR includes three numerical correctness bug fixes for the average pooling operator, which are important for our stakeholders working on safety-critical applications.

It’s been open for nearly a month, so I wanted to check in and see if there’s anything else needed from my side to move this forward.

Thank you very much!

**Comment by vivekkhandelwal1** [comment]
> Hi @vivekkhandelwal1,
> 
> Could you please take a look at this PR when you have a chance? I’ve addressed all the previous feedback, and GitHub requires your approval before it can be merged. This PR includes three numerical correctness bug fixes for the average pooling operator, which are important for our stakeholders working on safety-critical applications.
> 
> It’s been open for a month, so I wanted to check in and see if there’s anything else needed from my side to move this forward.
> 
> Thank you very much!

Hi @ivangarcia44, I'm sorry for the delay. I got caught up in some other work. Can I review it in a couple of days?

**Comment by ivangarcia44** [comment]
> > Hi @vivekkhandelwal1,
> > Could you please take a look at this PR when you have a chance? I’ve addressed all the previous feedback, and GitHub requires your approval before it can be merged. This PR includes three numerical correctness bug fixes for the average pooling operator, which are important for our stakeholders working on safety-critical applications.
> > It’s been open for a month, so I wanted to check in and see if there’s anything else needed from my side to move this forward.
> > Thank you very much!
> 
> Hi @ivangarcia44, I'm sorry for the delay. I got caught up in some other work. Can I review it in a couple of days?

Its ok, next week is perfectly fine. Thank you for following up!

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 10]
```diff
@@ -1040,27 +1061,59 @@ LogicalResult ConvertAtenAvgPoolOp<OpTy, PoolingOpTy, Dim>::matchAndRewrite(
...
+  // will give an extra element that will go out of bounds which PyTorch
+  // adds zero padding in it. It also does not count the implicit zero
+  // padding elements in the divisor, and it is not controlled by the
+  // count_include_pad argument.
+  // But also note that if all strides are 1 there are no fractions to
+  // round up, hence there is no ceiling rounding and the window will
+  // not go out of bounds. For this case the divisor is just the
+  // product of kernel dimensions.
+  // Search for torch.nn.AvgPool2d E2E tests for coverage of these
+  // conditions.
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:1100]: You may add something like this?
```suggestion
// Determines whether the average pooling divisor needs to be clamped
// (i.e., adjusted to exclude padded or out-of-bounds elements).
//
// There are two primary cases where clamping is needed:
// 1. Padding with count_include_pad == false:
//    - If padding is applied (padding != 0) and count_include_pad is false,
//      then padding elements are *excluded* from the divisor, effectively
//      clamping the divisor to the number of valid input elements.
//
// 2. Ceil mode with non-unit stride:
//    - When ceil_mode is enabled, output dimensions are rounded up, potentially
//      creating pooling windows that extend beyond the input tensor bounds.
//      PyTorch handles this by implicitly adding zero-padding outside the tensor,
//      but these extra (implicit) padded elements are *not* included in the divisor.
//      This behavior is independent of the count_include_pad flag.
//    - If all strides are 1, ceil_mode will not produce fractional divisions,
//      so the windows will not extend beyond bounds, and no clamping occurs.
//
// Reference: PyTorch AvgPool2d documentation and formula for H_out/W_out:
// https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html
//
// See torch.nn.AvgPool2d E2E tests for comprehensive coverage.
```

**Review by vivekkhandelwal1 — APPROVED** [review]
@ivangarcia44, thanks for the changes and for adding the clarifications. The PR looks good. If you want, then you may incorporate the comment suggestion before merging the PR. 

I have 2 doubts:

1) I am unable to understand the need for reversing the kernel/stride/padding element order.
2.) In response to this comment of mine, you said:
> Because of this I wrote a new average pooling divisor formula based on the PyTorch avg_pool2d doc and numerical values I observed.

Honestly, as per my opinion doing something based on observation won't be correct. There must be some rationale/reference for the same. Also. now I don't see that comment in the code.

**Comment by ivangarcia44** [comment]
> @ivangarcia44, thanks for the changes and for adding the clarifications. The PR looks good. If you want, then you may incorporate the comment suggestion before merging the PR.
> 
> I have 2 doubts:
> 
> 1. I am unable to understand the need for reversing the kernel/stride/padding element order.
> 2. In response to this comment of mine, you said:
> 
> > Because of this I wrote a new average pooling divisor formula based on the PyTorch avg_pool2d doc and numerical values I observed.
> 
> Honestly, as per my opinion doing something based on observation won't be correct. There must be some rationale/reference for the same. Also. now I don't see that comment in the code.

Hi Vivek, I will integrate your requested comment change. Here are my answers for the two questions above:
1) The issue is coming from upstream, the Pooling.cpp file is responsible for translating from torch to linalg. The spatial dimensions do not match the
